_posts/2018-05-20-Y-The-Most-Beautiful-Idea-in-Computer-Science.md

```diff
@@ -0,0 +1,611 @@
+author: Lucas Fernandes da Costa
+tags: functional-programming combinators javascript y-combinator
+---
+
+In this post we will talk about **one of the most beautiful ideas in computer science: the Y-Combinator**. And no, I'm not talking about [the VC firm in Sillicon Valley](http://www.ycombinator.com/), even though this post will explain why they've got this name.
+
+Put in simple terms, **the Y-Combinator (also known as the fixed-point combinator) is a way of doing recursion in a language that does not explicitly supports it**.
+
+Let's say you want to implement a recursive [factorial function](https://en.wikipedia.org/wiki/Factorial). In JavaScript, for example, you could simply do this:
+
+```js
+const factorial = n => n === 0 ? 1 : n * factorial(n - 1);
+```
+
+But what if you couldn't use any "names"? What if JavaScript didn't allow you to call `factorial` inside of `factorial`, how would you implement it?
+
+In that case, you'd use the Y-Combinator.
+
+I guarantee you **this will be one of the most amazing journeys you will go through in your life as a programmer**. The first time deriving the Y-Combinator is an indescribable joy, so get comfortable in your chair, bed, or whatever and let's get started.
+
+I recommend that you **follow this post with a notebook or on Node's REPL**, that will help you understand all the steps we'll go through and test things by yourself. **Don't be afraid of trying things out and testing things, doing your own exploration plays a great part in understanding how Y works**.
+
+Also, if you want to learn more about recursion [I've got a very popular post about it in this very same blog](/2017/05/08/All-About-Recursion-PTC-TCO-and-STC-in-JavaScript.html). It's a bit old but it remains amazing and very informative.
+
+
+## **First things first: what on earth is a combinator?**
+
+**A combinator is a function with no free variables**.
+
+Free variables are variables defined outside of the function's scope. They're the opposite of bound variables, which are variables that only exist inside of a function's scope.
+
+Let's use an example to make it easier to understand:
+
+```js
+const num = 13;
+
+const sumPlusThirteen = (a, b) => a + b + first
+```
+
+In the example above, our function `sumPlusThirteen` uses the variables `a`, `b` and `num`. **The variables `a` and `b` are bound variables because they are bound to the function's parameters**: when we refer to their names we are referring to values that were passed to the `sumPlusThirteen` function. However, **the variable `num` refers to a value outside of the function's scope, which means it is a free variable**.
+
+Let's look at some more examples:
+
+1. `x => x` - `x` is a bound variable, because it's bound to the function's parameters
+2. `x => x + y` - `x` is a bound variable because it's bound to the function's paramers and `y` is a free variable because it refers to a value outside of the function's scope
```

Comment:
You can just use `1.` for all the numbered items, and markdown renderer (at least for Jekyll and MkDocs) will automatically put the correct numbers in. This makes it easier to massage lists without having to manually rename things.

Reply:
That's amazing! I'll use it next time since this one is already done I'll keep it as is.